apps/hermes/server/src/state/aggregate.rs

```diff
@@ -568,6 +616,145 @@ where
+                        start_timestamp: start_twap.publish_time,
+                        end_timestamp: end_twap.publish_time,
+                        down_slots_ratio,
+                    });
+
+                    // Combine messages for update data
+                    let mut messages = Vec::new();
+                    messages.push(start_message.clone().into());
+                    messages.push(end_message.clone().into());
+
+                    if let Ok(update_data) = construct_update_data(messages) {
+                        all_update_data.push(update_data);
+                    } else {
+                        tracing::warn!(
+                            "Failed to construct update data for price feed {}",
+                            price_id
+                        );
+                        continue;
+                    }
+                }
+                Err(e) => {
+                    tracing::warn!(
+                        "Failed to calculate TWAP for price feed {}: {}",
+                        price_id,
+                        e
+                    );
+                    continue;
+                }
+            }
+        }
+    }
+
+    Ok(TwapsWithUpdateData {
+        twaps: calculated_twaps,
+        update_data: all_update_data,
+    })
+}
+fn calculate_twap(start_message: &TwapMessage, end_message: &TwapMessage) -> Result<Price> {
+    if end_message.publish_slot <= start_message.publish_slot {
+        return Err(anyhow!(
+            "Cannot calculate TWAP - end slot must be greater than start slot"
+        ));
+    }
+
+    // Validate that messages are the first ones in their timestamp
+    // This is necessary to ensure that this TWAP is deterministic,
+    // Since there can be multiple messages in a single second.
+    if start_message.prev_publish_time >= start_message.publish_time {
+        return Err(anyhow!(
+            "Start message is not the first update for its timestamp"
+        ));
+    }
+
+    if end_message.prev_publish_time >= end_message.publish_time {
+        return Err(anyhow!(
+            "End message is not the first update for its timestamp"
+        ));
+    }
+
+    let slot_diff = end_message
+        .publish_slot
+        .checked_sub(start_message.publish_slot)
+        .ok_or_else(|| anyhow!("Slot difference overflow"))?;
+
+    let price_diff = end_message
+        .cumulative_price
+        .checked_sub(start_message.cumulative_price)
+        .ok_or_else(|| anyhow!("Price difference overflow"))?;
+
+    let conf_diff = end_message
+        .cumulative_conf
+        .checked_sub(start_message.cumulative_conf)
+        .ok_or_else(|| anyhow!("Confidence difference overflow"))?;
+
+    // Perform division before casting to maintain precision
+    // Cast slot_diff to the same type as price / conf diff before division
+    let price = (price_diff / slot_diff as i128) as i64;
+    let conf = (conf_diff / slot_diff as u128) as u64;
```

Comment:
I think it's better to use the checked casts `u128::from` and `u64::try_from` generally

Reply:
oo great tip ty

Reply:
The `as` keyword for integer conversion is one the biggest mistakes in rust :D very un-rusty

Reply:
is there a rust equivalent to the [zen of python](https://peps.python.org/pep-0020/) out there? :)